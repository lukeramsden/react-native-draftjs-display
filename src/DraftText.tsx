import React from 'react';
import {Text, TextStyle, TextProps} from 'react-native';
import type {RawDraftContentBlock} from 'draft-js';
import {useDraftDisplayContext} from './DraftDisplay';
import {defaultBlockStyles, defaultInlineStyles} from './defaultStyles';
import {flattenInlineStyles} from './flattenInlineStyles';

export interface DraftTextProps {
  block: RawDraftContentBlock;
  TextComponent?: React.ComponentType<TextProps>;
}

const DefaultTextComponent = Text;

export const DraftText: React.FC<DraftTextProps> = ({
  block: {key, type, text, inlineStyleRanges, data},
  TextComponent = DefaultTextComponent,
}) => {
  const context = useDraftDisplayContext();

  const blockStyle =
    context.getCustomBlockStyles()?.[type as keyof typeof defaultBlockStyles] ||
    defaultBlockStyles[type as keyof typeof defaultBlockStyles] ||
    {};

  let inner: React.ReactNode;

  const flatInlineStyles = flattenInlineStyles(inlineStyleRanges);
  const textAlignStyle: {textAlign?: TextStyle['textAlign']} = {
    textAlign: (data as any)?.textAlign,
  };

  if (flatInlineStyles.length > 0) {
    // used to render any unstyled text before the start of the flattened styles
    let firstCharacterRendered: number = 0;
    // used to render any unstyled text after the end of the string
    let lastCharacterRendered: number = 0;

    inner = flatInlineStyles.map((inlineStyle, idx) => {
      if (idx === 0) {
        firstCharacterRendered = inlineStyle.offset;
      }
      lastCharacterRendered = inlineStyle.offset + inlineStyle.length;
      return (
        <TextComponent
          key={`${key}:${idx}`}
          style={[
            inlineStyle.style.map(
              (s) =>
                context.getCustomBlockStyles()?.[s] || defaultInlineStyles[s],
            ),
          ]}
          children={text.substr(inlineStyle.offset, inlineStyle.length)}
        />
      );
    });

    // add missing characters from the start
    if (firstCharacterRendered !== 0) {
      (inner as JSX.Element[]).unshift(
        <TextComponent
          key={`${key}:first`}
          children={text.substring(0, firstCharacterRendered)}
        />,
      );
    }

    // add rest of unstyled string
    if (text.length !== lastCharacterRendered) {
      (inner as JSX.Element[]).push(
        <TextComponent
          key={`${key}:last`}
          children={text.substring(lastCharacterRendered)}
        />,
      );
    }
  } else {
    inner = text;
  }

  return (
    <TextComponent
      style={[blockStyle, textAlignStyle.textAlign && textAlignStyle]}>
      {inner}
    </TextComponent>
  );
};

export default DraftText;
