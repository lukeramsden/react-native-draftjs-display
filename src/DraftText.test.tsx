import 'react-native';

import React from 'react';
import renderer from 'react-test-renderer';

import DraftText from './DraftText';
import type {RawDraftInlineStyleRange} from 'draft-js';

it('renders empty blocks', () => {
  const text = '';
  const tree = renderer
    .create(
      <DraftText
        block={{
          type: 'paragraph',
          text,
          inlineStyleRanges: [],
          entityRanges: [],
          key: 'test',
          depth: 0,
        }}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders text of unknown type omitting default styles', () => {
  const text = 'Hello World';
  const tree = renderer
    .create(
      <DraftText
        block={{
          type: 'unknowntype111222',
          text,
          inlineStyleRanges: [],
          entityRanges: [],
          key: 'test',
          depth: 0,
        }}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders unstyled text', () => {
  const text = 'Hello World';
  const tree = renderer
    .create(
      <DraftText
        block={{
          type: 'paragraph',
          text,
          inlineStyleRanges: [],
          entityRanges: [],
          key: 'test',
          depth: 0,
        }}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders styled text, single block', () => {
  const text = 'Hello World';
  const inlineStyleRanges: RawDraftInlineStyleRange[] = [
    {style: 'BOLD', offset: 0, length: 5},
  ];
  const tree = renderer
    .create(
      <DraftText
        block={{
          type: 'paragraph',
          text,
          inlineStyleRanges,
          entityRanges: [],
          key: 'test',
          depth: 0,
        }}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders styled text, overlapping', () => {
  const text = 'Hello World';
  const inlineStyleRanges: RawDraftInlineStyleRange[] = [
    {style: 'BOLD', offset: 0, length: 5},
    {style: 'ITALIC', offset: 4, length: 4},
  ];
  const tree = renderer
    .create(
      <DraftText
        block={{
          type: 'paragraph',
          text,
          inlineStyleRanges,
          entityRanges: [],
          key: 'test',
          depth: 0,
        }}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders text-align: left', () => {
  const text = 'Hello World';
  const data = {textAlign: 'left'};
  const tree = renderer
    .create(
      <DraftText
        block={{
          type: 'paragraph',
          text,
          data,
          inlineStyleRanges: [],
          entityRanges: [],
          key: 'test',
          depth: 0,
        }}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders text-align: right', () => {
  const text = 'Hello World';
  const data = {textAlign: 'right'};
  const tree = renderer
    .create(
      <DraftText
        block={{
          type: 'paragraph',
          text,
          data,
          inlineStyleRanges: [],
          entityRanges: [],
          key: 'test',
          depth: 0,
        }}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders text-align: center', () => {
  const text = 'Hello World';
  const data = {textAlign: 'center'};
  const tree = renderer
    .create(
      <DraftText
        block={{
          type: 'paragraph',
          text,
          data,
          inlineStyleRanges: [],
          entityRanges: [],
          key: 'test',
          depth: 0,
        }}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
