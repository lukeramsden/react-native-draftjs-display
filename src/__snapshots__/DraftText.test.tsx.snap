// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty blocks 1`] = `
<Text
  style={
    Array [
      Object {},
      undefined,
    ]
  }
>
  
</Text>
`;

exports[`renders styled text, overlapping 1`] = `
<Text
  style={
    Array [
      Object {},
      undefined,
    ]
  }
>
  <Text
    style={
      Array [
        Array [
          Object {
            "fontWeight": "bold",
          },
        ],
      ]
    }
  >
    Hell
  </Text>
  <Text
    style={
      Array [
        Array [
          Object {
            "fontWeight": "bold",
          },
          Object {
            "fontStyle": "italic",
          },
        ],
      ]
    }
  >
    o
  </Text>
  <Text
    style={
      Array [
        Array [
          Object {
            "fontStyle": "italic",
          },
        ],
      ]
    }
  >
     Wo
  </Text>
  <Text>
    rld
  </Text>
</Text>
`;

exports[`renders styled text, single block 1`] = `
<Text
  style={
    Array [
      Object {},
      undefined,
    ]
  }
>
  <Text
    style={
      Array [
        Array [
          Object {
            "fontWeight": "bold",
          },
        ],
      ]
    }
  >
    Hello
  </Text>
  <Text>
     World
  </Text>
</Text>
`;

exports[`renders text of unknown type omitting default styles 1`] = `
<Text
  style={
    Array [
      Object {},
      undefined,
    ]
  }
>
  Hello World
</Text>
`;

exports[`renders text-align: center 1`] = `
<Text
  style={
    Array [
      Object {},
      Object {
        "textAlign": "center",
      },
    ]
  }
>
  Hello World
</Text>
`;

exports[`renders text-align: left 1`] = `
<Text
  style={
    Array [
      Object {},
      Object {
        "textAlign": "left",
      },
    ]
  }
>
  Hello World
</Text>
`;

exports[`renders text-align: right 1`] = `
<Text
  style={
    Array [
      Object {},
      Object {
        "textAlign": "right",
      },
    ]
  }
>
  Hello World
</Text>
`;

exports[`renders unstyled text 1`] = `
<Text
  style={
    Array [
      Object {},
      undefined,
    ]
  }
>
  Hello World
</Text>
`;
